[RFC] Example use of NAMD CudaGlobalMaster interface

HanatoK opened the pull request:
Hi @giacomofiorin and @jhenin ! This is an example use of the NAMD's CudaGlobalMaster interface with Colvars. The data exchanged with CudaGlobalMaster are supposed to be GPU-resident. Due to Colvars' limitations the implementation still have to copy the data from GPU to CPU, but I think it is the first step to make Colvars GPU-resident (or at least a test-bed for the GPU porting).

## Compilation

This plugin depends on the source code of [NAMD](https://gitlab.com/tcbgUIUC/namd) (the [`opt_cudagm` branch](https://gitlab.com/tcbgUIUC/namd/-/tree/opt_cudagm?ref_type=heads)), [the CUDA compiler](https://developer.nvidia.com/cuda-downloads) and libraries and optionally the [TCL library](https://www.tcl-lang.org) (if your NAMD is built with TCL support). To compile this plugin you also need [cmake version 3.23 or above](https://cmake.org/download/).

Commands to compile the plugin under this directory:
```sh
mkdir build
cmake -DNAMD_DIR=<your_namd_git_directory>/ -DCMAKE_BUILD_TYPE=Release ../
make -j4
```
After the compilation, you will get a shared library named `libcudaglobalmastercolvars.so`.

### Notes for developers

If you built your NAMD with `-DNAMD_NVTX_ENABLED`, then you also need to switch that on in the CMake command by `-DNAMD_NVTX_ENABLED=on`. If you want to build the plugin on one computer but later want to run NAMD with it on another system, it is better to add `-DCMAKE_CUDA_ARCHITECTURES=all-major` to the CMake command.

## Example usage

This plugin is incompatible with the traditional way of using Colvars in NAMD, so you cannot use `colvars on` in your NAMD configuration and `numSteps` with it. Instead, please use the following TCL command in your NAMD configuration file before `run`:
```tcl
gpuGlobal on
gpuGlobalCreateClient <the_absolute_path_to_this_dir>/namd_cudaglobalmaster/build/libcudaglobalmastercolvars.so COLVARS <colvars_config_file>
```

The example NAMD input file can be found in `namd_cudaglobalmaster/example/alad.namd`, which dynamically loads the shared library built above and runs an OPES simulation along the two dihedral angles of the alanine dipeptide.

## Limitations

This plugin is still in its early stage, and has the following limitations:
- Limited scripting support. You can call Colvars between `run`s by `gpuGlobalUpdateClient COLVARS cv <command>`. For example, you can reset Colvars by `gpuGlobalUpdateClient COLVARS cv reset`, and then load another configuration file by `gpuGlobalUpdateClient COLVARS cv configfile <new_config_file>`. However, you cannot use scripted Colvars force, and since this plugin uses an independent interpreter for the Colvars instance, so you cannot call NAMD's TCL procedures in the calculation of scripted CVs;
- `volmap` is not available;
- GaMD energy histograming/reweighting is not available (GaMD is not currently available in GPU-resident NAMD);
- SMP is not available.

HanatoK commented:
Another issue:

It is a bit strange that this plugin of CudaGlobalMaster loads the `colvarproxy` related symbols from NAMD, instead of the code Colvars source compiled with it.

HanatoK commented:
> With the optimization in https://gitlab.com/tcbgUIUC/namd/-/merge_requests/402 and the fix of dlopen in https://gitlab.com/tcbgUIUC/namd/-/merge_requests/400, I have carried out another benchmark using the RMSD test case that @giacomofiorin sent to me. The system has 315,640 atoms, and the RMSD CV involves 41,836 atoms. The simulations were run on AMD Ryzen Threadripper PRO 3975WX + NVIDIA RTX A6000.
> 
> Simulation	Speed (ns/day)
> No Colvars	32.6442
> Colvars with this interface	27.1574
> Colvars with GlobalMaster	14.2476

The benchmark above does not enable PME. I have implemented CUDA allocator for the vectors in Colvars (also with some optimizations on the NAMD side), and conducted the benchmark again with PME and on more platforms:

**AMD Ryzen Threadripper PRO 3975WX + NVIDIA RTX A6000**

|Simulation| Speed (ns/day)|
|---------------|-----------------------|
| No Colvars| 27.8499|
|Colvars with CudaGlobalMaster| 26.871|
|Colvars with GlobalMaster| 13.9529|

**Intel(R) Xeon(R) Platinum 8168 CPU + 2xTesla V100-SXM3-32GB**

|Simulation| Speed (ns/day)|
|---------------|-----------------------|
| No Colvars| 54.1266|
|Colvars with CudaGlobalMaster| 38.1852|
|Colvars with GlobalMaster| 7.22716|

**ARM Neoverse-V2 +  NVIDIA GH200**

|Simulation| Speed (ns/day)|
|---------------|-----------------------|
| No Colvars| 78.6648|
|Colvars with CudaGlobalMaster| 67.9719|
|Colvars with GlobalMaster| 22.5077|

jhenin commented:
Impressive results! Do you see a path to making this work transparently in specialized, static builds of Colvars/NAMD, or are there lasting reasons that this needs to be a dynamic library?

HanatoK commented:
> Impressive results! Do you see a path to making this work transparently in specialized, static builds of Colvars/NAMD, or are there lasting reasons that this needs to be a dynamic library?

No. This implementation uses a new interface (I call it CudaGlobalMaster) which is independent of the old GlobalMaster interface. The GlobalMaster interface was designed more than twenty years ago when there was no GPU computing. To be compatible with GlobalMaster, the GPU-resident code path has to copy all atoms from the GPU memory to SOA buffers of patches, and then the `ComputeGlobal` objects convert them to AOS buffers, and then the `ComputeGlobal` objects send the aggregated atoms to `GlobalMasterServer` via Charm++ message, and finally `GlobalMasterServer` copies the requested atoms to Colvars (as a client derived from `GlobalMaster`). As you can see, the data copying is massive and very indirect. I am not sure how to improve it without breaking many things like the CPU MPI build.

`CudaGlobalMaster` is specialized for the GPU-resident mode, only copies those atoms that are requested by clients and copies them only once (for Colvars it has to be twice because Colvars requires an extra GPU-CPU data copying). Dynamic loading is more flexible because the plugins can link to any other 3rd libraries as they want. If I use dynamic/static linking for the interface, then the clients that use pytorch or tensorflow would force NAMD to link them to, which makes NAMD nearly impossible to distribute.

giacomofiorin commented:
I totally agree with @jhenin!

Many of us thought that much of the slowdown when running NAMD GPU-resident + Colvars came from the way data is copied (which is inherited by the constraints of GlobalMaster, an almost *30*-year old piece of code). But I am still amazed that you managed to get such a speedup without even touching the (slow and inefficient) code in the Colvars library. Absolutely impressive indeed!

If you think that dynamic linkage is absolutely required, we could work with it: you just demonstrated very clearly the value of supporting CUDAGlobalMaster.

That said, in making your considerations please also factor in that non-static executables take more work to install and maintain, either by the users or by their support staff. Academic institutions have historically had a difficult time recruiting good sysadmins, and (at least in the US) this is getting even harder lately :-(

HanatoK commented:
> Thanks for the explanations @HanatoK . I'm sure sysadmins would manage to build this if they had to, but if there is any way at all we can maintain the out-of-the-box, seamless Colvars experience, I think that would be a huge benefit to our users. Right now they have inputs that run seamlessly with GlobalMaster. If they could use the same input and the official binary from Illinois and get the performance that you just unlocked, that would be just awesome. I'm willing to put my own time and effort into this if there is any chance to make it happen.

Thanks! This interface is still preliminary. I just completed the TCL integration partially today and I will need to test it. I still don't know how `colvarscript` works, and I don't even know whether `colvarscript` requires TCL or not. From the CudaGlobalMaster interface side, I have implemented `virtual std::string updateFromTCLCommand(const std::vector<std::string>& arguments)` that accepts arguments between `run`s and allows the client to do anything it wants. In other words, I expect that Colvars has a general scripting interface like `int cvscript_run(int argc, char* argv[])`.

giacomofiorin commented:
> > I'm willing to put my own time and effort into this if there is any chance to make it happen.

Likewise from me :-)

Having Colvars in all official builds reaches a fairly large user base, which has evolved to rely on it (often in not so visible ways). Besides the several tutorials that exist around, CHARMM-GUI implicitly relies on Colvars when producing NAMD input decks for most membrane systems. There is a high chance that a typical NAMD user will also become a Colvars user at some point.

> Thanks! This interface is still preliminary. I just completed the TCL integration partially today and I will need to test it. I still don't know how `colvarscript` works, and I don't even know whether `colvarscript` requires TCL or not. From the CudaGlobalMaster interface side, I have implemented `virtual std::string updateFromTCLCommand(const std::vector<std::string>& arguments)` that accepts arguments between `run`s and allows the client to do anything it wants. In other words, I expect that Colvars has a general scripting interface like `int cvscript_run(int argc, char* argv[])`.

Here it looks like something we talked about earlier could help. Contrary to the early days, it makes less and less sense to derive `colvarproxy_namd` from `GlobalMaster`. Even for a plugin implementation, it would be better to consolidate what is common as much as possible, have a more abstract interface that supports both `GlobalMaster` and `CUDAGlobalMaster` (based on the availability of the code and the user's input).

HanatoK commented:
> Here it looks like something we talked about earlier could help. Contrary to the early days, it makes less and less sense to derive `colvarproxy_namd` from `GlobalMaster`. Even for a plugin implementation, it would be better to consolidate what is common as much as possible, have a more abstract interface that supports both `GlobalMaster` and `CUDAGlobalMaster` (based on the availability of the code and the user's input).

Again, I am afraid that since `CUDAGlobalMaster` greatly differs from `GlobalMaster`, it is not very meaningful to have an abstract interface. `CUDAGlobalMaster` does not support SMP, and it calls clients on a specific PE (master PE) that controls the GPU device, which may not be PE 0. The scripting interface (`ScriptTcl`) of NAMD only supports PE 0. To support scripting of any clients, I have added `ScriptTcl::Tcl_gpuGlobalUpdateClient` that basically broadcasts the TCL arguments to all PEs, and only the master PE will receive the arguments. More specifically, in `namd_cudaglobalmaster/example/alad.namd` of this PR, the line
```tcl
gpuGlobalCreateClient ../build/libcudaglobalmastercolvars.so COLVARS opes.colvars
```
creates a client instance `COLVARS`, and the NAMD TCL command
```tcl
set result [gpuGlobalUpdateClient COLVARS xxx yyy zzz]
```
will pass the `xxx yyy zzz` as a `std::vector<std::string>` to the `COLVARS` instance, and expect an `std::string` result.

I think when we say Colvars scripting there are essentially two "directions":
1. Colvars calls some scripts to compute the CVs and biases. I guess this can be solved by simply using `set_tcl_interp` 
2. A scripting language calls Colvars. In such case it seems calling `proxy->script->run` is enough but I am not sure, and this is not how the traditional `GlobalMaster` interface work.

The only shared code for both interfaces seem to be updating of masses and charges, PDB readers, setting of simulation temperature and I/O streams.

giacomofiorin commented:
> Again, I am afraid that since `CUDAGlobalMaster` greatly differs from `GlobalMaster`, it is not very meaningful to have an abstract interface. 

That is absolutely true: I most certainly do not want to suggest that those two wildly different classes should have a shared API :smile:

jhenin commented:
Then, if we agree on removing the inheritance, could `colvarproxy_namd` have a `GlobalMaster` and a `CUDAGlobalMaster` member, and switch between them as appropriate?

HanatoK commented:
> Then, if we agree on removing the inheritance, could `colvarproxy_namd` have a `GlobalMaster` and a `CUDAGlobalMaster` member, and switch between them as appropriate?

I am not sure how removing the inheritance works. There still should be a class derived from `GlobalMaster` to make Colvars work with the `GlobalMaster` interface. Could you tell me more details about your plan?

giacomofiorin commented:
> I am not sure how removing the inheritance works. There still should be a class derived from `GlobalMaster` to make Colvars work with the `GlobalMaster` interface. Could you tell me more details about your plan?

[This branch](https://github.com/Colvars/colvars/tree/globalmaster_decouple) contains a commit that makes `GlobalMasterColvars` a real class that implements a thin wrapper around `colvarproxy_namd` (as opposed to the latter inheriting from the former).

It passes all tests, minus the ones related to the volmaps: if I can't fix those with a bit more time, it's not a deal breaker for me because my preference would be discontinuing that code path altogether (in #737 and later work).

HanatoK commented:
@giacomofiorin I have tried to implement the scripting in the interface of CudaGlobalMaster in the https://github.com/Colvars/colvars/pull/783/commits/66ed6f6fb0bd74dd26e15d95e11cb5a7451c6946 commit. I mainly followed your LAMMPS interface code, and it seems working for commands like `cv getnumatoms`. But `cv reset`  seems broken. What do I need to do to support `cv reset`? Could you help me take a look at the code if you have time?

giacomofiorin commented:
> @giacomofiorin I have tried to implement the scripting in the interface of CudaGlobalMaster in the [66ed6f6](https://github.com/Colvars/colvars/commit/66ed6f6fb0bd74dd26e15d95e11cb5a7451c6946) commit. I mainly followed your LAMMPS interface code, and it seems working for commands like `cv getnumatoms`. But `cv reset` seems broken. What do I need to do to support `cv reset`? Could you help me take a look at the code if you have time?

Absolutely! Reaching out via chat.

HanatoK commented:
I marked this as ready. The updated instruction of compilation and usage can be found in the [README.md](https://github.com/HanatoK/colvars/blob/api_cudagm/namd_cudaglobalmaster/README.md). I think the only change to the source code of Colvars itself is the introduction of custom allocator that creates `std::vector` on host-pinned memory.

giacomofiorin commented:
Hi @HanatoK I found a minor issue (need to include the Charm++ header directory, easily fixable), but I am not sure how to handle this.
```
In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:2:
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:14:32: error: ‘CudaGlobalMaster’ does not name a type; did you mean ‘CudaGlobalMasterServer’?
   14 | using CudaGlobalMasterClient = CudaGlobalMaster::CudaGlobalMasterClient;
      |                                ^~~~~~~~~~~~~~~~
      |                                CudaGlobalMasterServer
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:22:7: error: conflicting return type specified for ‘virtual int CudaGlobalMasterColvars::updateFromTCLCommand(const std::vector<std::__cxx11::basic_string<char> >&)’
   22 |   int updateFromTCLCommand(const std::vector<std::string>& arguments) override;
      |       ^~~~~~~~~~~~~~~~~~~~
In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:1:
/home/fioring2/sources/namd/main/src/CudaGlobalMasterClient.h:294:23: note: overridden function is ‘virtual std::string CudaGlobalMasterClient::updateFromTCLCommand(const std::vector<std::__cxx11::basic_string<char> >&)’
  294 |   virtual std::string updateFromTCLCommand(const std::vector<std::string>& arguments);
      |                       ^~~~~~~~~~~~~~~~~~~~
In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:2:
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:23:8: error: ‘void CudaGlobalMasterColvars::onBuffersUpdated()’ marked ‘override’, but does not override
   23 |   void onBuffersUpdated() override;
      |        ^~~~~~~~~~~~~~~~
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:67:15: error: ‘std::string CudaGlobalMasterColvars::getTCLUpdateResult()’ marked ‘override’, but does not override
   67 |   std::string getTCLUpdateResult() override {return mTCLResult;}
      |               ^~~~~~~~~~~~~~~~~~
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C: In constructor ‘CudaGlobalMasterColvars::CudaGlobalMasterColvars()’:
/home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:965:29: error: ‘getScript’ is not a member of ‘CudaGlobalMasterClient’
  965 |     CudaGlobalMasterClient::getScript());
      |                             ^~~~~~~~~
```

I am using the same CentOS9-devel container that we use for the CI jobs, plus a CUDA 11.8 local installation loaded in the $PATH variable.

HanatoK commented:
> Hi @HanatoK I found a minor issue (need to include the Charm++ header directory, easily fixable), but I am not sure how to handle this.
> 
> ```
> In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:2:
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:14:32: error: ‘CudaGlobalMaster’ does not name a type; did you mean ‘CudaGlobalMasterServer’?
>    14 | using CudaGlobalMasterClient = CudaGlobalMaster::CudaGlobalMasterClient;
>       |                                ^~~~~~~~~~~~~~~~
>       |                                CudaGlobalMasterServer
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:22:7: error: conflicting return type specified for ‘virtual int CudaGlobalMasterColvars::updateFromTCLCommand(const std::vector<std::__cxx11::basic_string<char> >&)’
>    22 |   int updateFromTCLCommand(const std::vector<std::string>& arguments) override;
>       |       ^~~~~~~~~~~~~~~~~~~~
> In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:1:
> /home/fioring2/sources/namd/main/src/CudaGlobalMasterClient.h:294:23: note: overridden function is ‘virtual std::string CudaGlobalMasterClient::updateFromTCLCommand(const std::vector<std::__cxx11::basic_string<char> >&)’
>   294 |   virtual std::string updateFromTCLCommand(const std::vector<std::string>& arguments);
>       |                       ^~~~~~~~~~~~~~~~~~~~
> In file included from /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:2:
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:23:8: error: ‘void CudaGlobalMasterColvars::onBuffersUpdated()’ marked ‘override’, but does not override
>    23 |   void onBuffersUpdated() override;
>       |        ^~~~~~~~~~~~~~~~
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.h:67:15: error: ‘std::string CudaGlobalMasterColvars::getTCLUpdateResult()’ marked ‘override’, but does not override
>    67 |   std::string getTCLUpdateResult() override {return mTCLResult;}
>       |               ^~~~~~~~~~~~~~~~~~
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C: In constructor ‘CudaGlobalMasterColvars::CudaGlobalMasterColvars()’:
> /home/fioring2/sources/colvars/api_cudagm/namd_cudaglobalmaster/colvarproxy_cudaglobalmaster.C:965:29: error: ‘getScript’ is not a member of ‘CudaGlobalMasterClient’
>   965 |     CudaGlobalMasterClient::getScript());
>       |                             ^~~~~~~~~
> ```
> 
> I am using the same CentOS9-devel container that we use for the CI jobs, plus a CUDA 11.8 local installation loaded in the $PATH variable.

This plugin relies on the NAMD branch that hasn't been merged yet. For the time being you need to checkout the NAMD MR in https://gitlab.com/tcbgUIUC/namd/-/merge_requests/402 or the https://gitlab.com/tcbgUIUC/namd/-/tree/opt_cudagm branch.

giacomofiorin commented:
> @giacomofiorin The previous PR message is a bit outdated. I have updated it according to https://github.com/HanatoK/colvars/blob/api_cudagm/namd_cudaglobalmaster/README.md

Ok thanks, it built no

giacomofiorin commented:
I can confirm the excellent performance using my own test run on a local computer.  Using a Xeon w7-2475X CPU and a RTX 4500 Ada gen GPU, I get the following results (formatted as above):

**Simulation** | **Speed (ns/day)**
-- | --
No Colvars | 29.110
Colvars with CudaGlobalMaster | 28.616
Colvars with GlobalMaster | 16.873

giacomofiorin reviewed (approved):
As noted in the comments, the performance gain for this alternative interface between Colvars and NAMD GPU-resident is quite impressive. It is also relatively straightforward to [reproduce](https://github.com/Colvars/colvars/pull/783#issuecomment-2798072399).

There are some minor usability limitations (noted in the README) but those can definitely be sorted out later.

The only two issues worth mentioning here are:
1. The extensive duplication of the existing GlobalMaster-based interface has the potential to create diverging histories. Perhaps this can be reduced by separating out the communication scheme (GlobalMaster or CUDAGlobalMaster) from the rest of the NAMD interface (see [here](https://github.com/Colvars/colvars/pull/783#issuecomment-2759461186)).  EDIT: IMO does not prevent merging, though.
2. The addition of a top-level folder `namd_cudaglobalmaster` is in apparent contrast with #676. @jhenin Please comment?

Other than that, this looks like a significant breakthrough, paving the way for much better GPU support.

HanatoK commented on src/colvarproxy.h:
```diff
@@ -44,6 +44,16 @@ class colvarproxy_atoms {
+  using allocator_type = std::allocator<T>;
+#endif
+  using real_vector_t = std::vector<cvm::real, allocator_type<cvm::real>>;
+  using rvector_vector_t = std::vector<cvm::rvector, allocator_type<cvm::rvector>>;
```
Done.

giacomofiorin commented on namd_cudaglobalmaster/.gitignore:
```diff
@@ -0,0 +1,3 @@
+/build/*
+/example/output/*
+.clangd
```
No, of course not. I'm referring to `.clangd` only

jhenin commented:
As @giacomofiorin said, I would like to reduce the number of top-level subdirectories in this repo. For example, could `namd_cudaglobalmaster` move to `namd/cudaglobalmaster` ?

jhenin reviewed (approved):
This is a huge performance improvement, I have no technical objections on the core. Since this depends on an unmerged branch of NAMD and requires separate compilation, it will be limited to power users for the time being, but that does not preclude merging: the interface can be made more easily accessible later. My only request is not to clutter the main directory of the repo by moving files to a subdirectory of `namd`.

HanatoK commented:
> This is a huge performance improvement, I have no technical objections on the core. Since this depends on an unmerged branch of NAMD and requires separate compilation, it will be limited to power users for the time being, but that does not preclude merging: the interface can be made more easily accessible later. My only request is not to clutter the main directory of the repo by moving files to a subdirectory of `namd`.

Thank you for your review! I have moved the source code directory into `namd/`. Since there are test failures I have also rebased the branch on `master`.

giacomofiorin commented:
@HanatoK Where do you think the folder would be for the new files in the NAMD repository? Should they be under `src/` (like the current `colvarproxy_namd.C` or in a different folder?

HanatoK commented:
> @HanatoK Where do you think the eventual location would be for the new files in the NAMD repository? Should they be under `src/` (like the current `colvarproxy_namd.C` or in a different folder?

I think these new files can be maintained out of the NAMD repository so basically they can be anywhere in the Colvars repo, but since they are different from the traditional GlobalMaster code path, I don't think putting them in `src/` is a good idea.

giacomofiorin commented:
> I think these new files can be maintained out of the NAMD repository so basically they can be anywhere in the Colvars repo, but since they are different from the traditional GlobalMaster code path, I don't think putting them in `src/` is a good idea.

Sounds good, so the folder `cudaglobalmaster` could maybe work as a top-level folder in the NAMD repo, if these files eventually go in there as well.
